src/analyser.cpp

```diff
@@ -0,0 +1,1523 @@
+        && (astGreatGrandParent != nullptr) && (astGreatGrandParent->mPimpl->mType == AnalyserEquationAst::Type::DIFF)) {
+        double value;
+
+        if (!convertToDouble(ast->mPimpl->mValue, value) || !areEqual(value, 1.0)) {
+            auto issue = Issue::create();
+            auto variable = astGreatGrandParent->mPimpl->mRightChild->mPimpl->mVariable;
+
+            issue->setDescription("The differential equation for variable '" + variable->name()
+                                  + "' in component '" + owningComponent(variable)->name()
+                                  + "' must be of the first order.");
+            issue->setCause(Issue::Cause::MATHML);
+
+            mAnalyser->addIssue(issue);
+        }
+    }
+
+    // Make a variable a state if it is used in an ODE.
+
+    if ((ast->mPimpl->mType == AnalyserEquationAst::Type::CI)
+        && (astParent != nullptr) && (astParent->mPimpl->mType == AnalyserEquationAst::Type::DIFF)) {
+        internalVariable(ast->mPimpl->mVariable)->makeState();
+    }
+
+    // Recursively check the given AST's children.
+
+    if (ast->mPimpl->mLeftChild != nullptr) {
+        analyseEquationAst(ast->mPimpl->mLeftChild);
+    }
+
+    if (ast->mPimpl->mRightChild != nullptr) {
+        analyseEquationAst(ast->mPimpl->mRightChild);
+    }
+}
+
+double Analyser::AnalyserImpl::scalingFactor(const VariablePtr &variable)
+{
+    // Return the scaling factor for the given variable.
+
+    return Units::scalingFactor(variable->units(), internalVariable(variable)->mVariable->units());
+}
+
+void Analyser::AnalyserImpl::scaleAst(const AnalyserEquationAstPtr &ast,
+                                      const AnalyserEquationAstPtr &astParent,
+                                      double scalingFactor)
+{
+    // Scale the given AST using the given scaling factor.
+
+    auto scaledAst = AnalyserEquationAst::create();
+
+    scaledAst->mPimpl->populate(AnalyserEquationAst::Type::TIMES, astParent);
+
+    scaledAst->mPimpl->mLeftChild = AnalyserEquationAst::create();
+    scaledAst->mPimpl->mRightChild = ast;
+
+    scaledAst->mPimpl->mLeftChild->mPimpl->populate(AnalyserEquationAst::Type::CN, convertToString(scalingFactor), scaledAst);
+
+    ast->mPimpl->mParent = scaledAst;
+
+    if (astParent->mPimpl->mLeftChild == ast) {
+        astParent->mPimpl->mLeftChild = scaledAst;
+    } else {
+        astParent->mPimpl->mRightChild = scaledAst;
+    }
+}
+
+void Analyser::AnalyserImpl::scaleEquationAst(const AnalyserEquationAstPtr &ast)
+{
+    // Recursively scale the given AST's children.
+
+    if (ast->mPimpl->mLeftChild != nullptr) {
+        scaleEquationAst(ast->mPimpl->mLeftChild);
+    }
+
+    if (ast->mPimpl->mRightChild != nullptr) {
+        scaleEquationAst(ast->mPimpl->mRightChild);
+    }
+
+    // If the given AST node is a variable (i.e. a CI node) then we may need to
+    // do some scaling.
+
+    if (ast->mPimpl->mType == AnalyserEquationAst::Type::CI) {
```

Comment:
No, we don't scale. The only scaling that is done is at the connection level. That's what the CellML spec says and requires.

This is also the reason @awickens has been working on units within equations, so that we can generate warnings/hints as needed.